model/dyn_core.F90

```diff
@@ -1370,6 +1392,25 @@ subroutine dyn_core(npx, npy, npz, ng, sphum, nq, bdt, n_map, n_split, zvir, cp,
   enddo   ! time split loop
   first_call=.false.
 !-----------------------------------------------------
+!The following is for SA-3D-TKE (kyf) (modify for data structure)
+!--calculating shear deformation and TKE transport for 3d TKE scheme
+    if(sa3dtke_dyco) then
+     if ( end_step ) then
+       ntke = get_tracer_index(MODEL_ATMOS, 'sgs_tke')
+       call mpp_update_domains(q(:,:,:,ntke), domain)
+       call mpp_update_domains(sa3dtke_var%dku3d_e(:,:,:), domain) ! update dku3d_e at halo
```

Comment:
These are inefficient halo updates as two MPI communications are done - one for each variable. I suggest using the optional COMPLETE argument with proper flags to buffer the updates and have them occur in a single communication.